src/primitives/hrpstring.rs

```diff
@@ -0,0 +1,447 @@
+        if ret.data.len() * 5 % 8 > 4 {
+            return Err(Error::InvalidDataLength);
+        }
+
+        Ok((ret, witver))
+    }
+
+    /// Returns the witness version if present.
+    pub fn witness_version(&self) -> Option<u8> { self.witness_version }
+
+    /// Returns the human-readable part.
+    pub fn hrp(&self) -> Hrp { self.hrp }
+
+    /// Returns an iterator over the byte data encoded by the HRP string (excluding the HRP, witness
+    /// version byte if any, and checksum).
+    pub fn byte_iter(&self) -> ByteIter {
+        ByteIter { iter: AsciiToFe32Iter { iter: self.data.iter().copied() }.fe32_to_byte_iter() }
+    }
+
+    /// Returns an iterate over field elements of the data encoded by the HRP string (excluding the
+    /// HRP, witness version byte if any, and checksum).
+    pub fn fe32_iter(&self) -> Fe32Iter {
+        Fe32Iter { iter: AsciiToFe32Iter { iter: self.data.iter().copied() } }
+    }
+
+    /// Parses an bech32 encode string and constructs a [`Parsed`] object that must have
+    /// `validate_checksum` called (if it contains a checksum).
+    ///
+    /// Checks for valid ASCII values, does not validate the checksum.
+    fn parse_unvalidated(s: &'s str) -> Result<Self, Error> {
+        let sep_pos = check_characters(s)?;
+        let (hrp, data) = s.split_at(sep_pos);
+
+        let ret = Parsed {
+            hrp: Hrp::parse(hrp)?,
+            witness_version: None,
+            data: data[1..].as_bytes(), // Skip the separator.
+        };
+
+        Ok(ret)
+    }
+
+    /// Validates that a [`Parsed`] returned by `parse_unvalidated` has a valid checksum.
+    ///
+    /// # Returns
+    ///
+    /// Returns `self` with the checksum removed from the inner data slice.
+    fn validate_checksum<Ck: Checksum>(mut self) -> Result<Self, Error> {
+        if Ck::CHECKSUM_LENGTH == 0 {
+            return Ok(self); // Called with NoChecksum.
+        }
+
+        if self.data.len() < Ck::CHECKSUM_LENGTH {
+            return Err(Error::InvalidChecksumLength);
+        }
+
+        let mut checksum_eng = checksum::Engine::<Ck>::new();
+        checksum_eng.input_hrp(&self.hrp());
+        if let Some(witver) = self.witness_version {
+            checksum_eng.input_fe(Fe32::try_from(witver).map_err(Error::InvalidWitnessVersion)?);
+        }
+        // Unwrap ok since we checked all characters in our constructor.
+        for fe in self.data.iter().map(|&b| Fe32::from_char_unchecked(b)) {
+            checksum_eng.input_fe(fe);
+        }
+
+        if checksum_eng.residue() != &Ck::TARGET_RESIDUE {
+            return Err(Error::InvalidChecksum);
+        }
+
+        let data_len = self.data.len() - Ck::CHECKSUM_LENGTH;
+        self.data = &self.data[..data_len];
+
+        Ok(self)
+    }
+}
+
+/// A iterator over a parsed HRP string data as bytes.
+#[allow(clippy::type_complexity)]
+pub struct ByteIter<'s> {
+    // We need `Copied` because non of our adaptor iterators use referenced items.
```

Comment:
In e2527dc19c3e8e9a4cd678a955fc5b161e175c87:

typo `non`. TBH you could probably delete the comment entirely, the purpose of `Copied` is pretty clear.

Also, I think `allow(clippy::type_complexity)` is probably the right approach here but FYI you can edit `type-complexity-threshold` in clippy.toml to just increase its threshold. The default is 250. In my experience 1000 shuts it up (and if it doesn't maybe you **should** break up the types :P).

Reply:
Nice, used 1000 as suggested.